### ISSUE TITLE
Paginated methods lose typing information for type-checking

### ISSUE THREAD

**Issue by sirosen** [globus/globus-sdk-python#490]
```python
import globus_sdk

reveal_type(globus_sdk.TransferClient().paginated.task_event_list)
```
shows
`Revealed type is "def (*Any, **Any) -> Any"`

This is because the type found for attributes of `paginated` comes from `__getattr__` on the paginator table, which is annotated as `-> Callable` (because the precise type depends on the name being used).

mypy issues details why this is complicated, e.g. https://github.com/python/mypy/issues/5523
I don't think we can expect `__getattr__` to solve this correctly for us.

I tried some trivial hacks, but nothing is easily working for this case.
A paginator table is a partial proxy for a client, so there's always the possibility of casting it to the relevant client class. Although odd, it would get you the right signature on the resulting function.

**Comment by sirosen** [comment]
Just to record it, I did try putting together a mypy plugin to see if it's possible to get paginator types with a plugin.
I'd like it -- even if it's experimental -- for use in the CLI. The following plugin code _doesn't_ work because I wasn't able to solve getting the attribute name from the AttributeContext:
```python
"""
Experimental: a mypy plugin for special types within the Globus SDK
"""

import typing as t

from mypy.plugin import AttributeContext, Plugin
from mypy.types import Type, get_proper_type


def _paginated_method_callback(ctx: AttributeContext) -> Type:
    # do not modify the type if the object type is not generic
    if len(ctx.type.args) != 1:
        return ctx.default_attr_type
    # if there is no name for this node, do not change the type
    if ctx.type.fullanme is None:
        return ctx.default_attr_type

    client_class_type = get_proper_type(ctx.type.args[0])
    attribute_name = ...  # UNKNOWN: how do we get the attribute name in this context?

    # TODO: this returns the underlying client method type; should transform to a
    # paginator as the result type
    return get_proper_type(getattr(client_class_type, attribute_name))


class GlobusSDKPlugin(Plugin):
    def get_attribute_hook(
        self, fullname: str
    ) -> t.Optional[t.Callable[[AttributeContext], Type]]:
        if fullname.startswith("globus_sdk.paging.table.PaginatorTable"):
            return _paginated_method_callback
        return None


def plugin(version: str) -> t.Type[Plugin]:
    return GlobusSDKPlugin
```

**Comment by sirosen** [comment]
Closing as resolved by #494
